### DISCUSSION TITLE
Writing language pack for atom

### DISCUSSION THREAD

**Discussion by GordianDziwis** [tree-sitter/tree-sitter#1213] [Q&A]
I wrote parser for sparql and [turtle](https://github.com/BonaBeavis/tree-sitter-turtle), and like to add them to atom.
I have already added them to [neovim](https://github.com/nvim-treesitter/nvim-treesitter) and because of this I build the parser with treesitter 19.2.

But Atom depends on treesitter 17.2, so I cant use my parser because of ABI version missmatch.
A [pull request](https://github.com/atom/atom/pull/22130) for Atom went stale.

Can I build my parser somehow for different ABI versions? Is there another way I did overlook?

**Comment by bookofproofs** [comment]
This would explain, why I'm also still having problems with getting my Atom language package `language-fpl` running. I created my parser with the newest tree-sitter and have been trying for two weeks now to get my Atom language package working - with no success so far. 

Thank you for your question :-). Unfortunately, I do not know the answer. I'll passively follow your discussion because maybe the answer is what I'm also looking for.

**Reply by GordianDziwis** [reply]
You can validate if you have the same issue by open a file with your grammar activated. And opening the developer mode in atom and check for an error in the console containing something with "RangeError: Incompatible language version. Compatible range: 9 - 12. Got: 13"

**Reply by bookofproofs** [reply]
Did you mean writing in the console `atom --dev`? I did it and I opened an Atom window in dev mode. But the console does not report any errors. Sorry, I'm not familiar with Atom.

**Reply by tmatijevich** [reply]
`atom -d`
Then when launched `ctrl-shift-i` for the [Developer's Console](https://flight-manual.atom.io/hacking-atom/sections/debugging/#check-for-errors-in-the-developer-tools)

**Reply by bookofproofs** [reply]
gotcha: `Uncaught (in promise) RangeError: Incompatible language version. Compatible range: 9 - 12. Got: 13`

**Reply by tmatijevich** [reply]
- [Aerijo's guide](https://gist.github.com/Aerijo/df27228d70c633e088b0591b8857eeef) has helped when working with tree-sitter and Atom. (Python 2.7 was not a requirement for me)
    - This suggests `nvm` to overcome the version mismatches
- @bookofproofs looks like you're on Windows, so you may want [nvm-windows](https://github.com/coreybutler/nvm-windows#readme)
    - I would definitely try the atom grammar package and tree-sitter package both [locally](https://github.com/Xapphire13/tree-sitter-syntax-visualizer/issues/11#issuecomment-872193854) before publishing
- @BonaBeavis [this answer](https://stackoverflow.com/a/64557291/11402416) also worked for me when there was an electron version mismatch, but `nvm` may be the better solution

**Reply by bookofproofs** [reply]
@tmatijevich : Thank you for your link to Aeirjo's guide. I was looking for something like this. I manage to test the tree-sitter package locally, but I do not know how to do it for Atom packages. For instance, how can I force Atom to open a package that is not yet published and that still resides on my local Atop package repository I'm working on? 

I wished I could see the behavior of Atom with a still unpublished package so I can correct it before publishing. But I do not know how to do it. As I already said, I'm not familiar with Atom.

**Reply by tmatijevich** [reply]
I [commented here](https://github.com/Xapphire13/tree-sitter-syntax-visualizer/issues/11#issuecomment-873258326) and [here's related documentation](https://flight-manual.atom.io/hacking-atom/sections/hacking-on-atom-core/).

**Reply by GordianDziwis** [reply]
@tmatijevich The issue is not that there is a node or electron version mismatch. The mismatch is between the treesitter version packaged with atom and the treesitter version the grammar's parser was build [with](https://github.com/BonaBeavis/tree-sitter-turtle/blob/ae634171e8ea4fd88fe3199bc27d03abf2d4c7e4/src/parser.c#L8).

**Comment by claytonrcarter** [comment]
I'm also fumbling around this and found that there is a `--prev-abi` flag for the `generate` command. For me, running `tree-sitter generate` will generate for ABI 13, while`tree-sitter generate --prev-abi` produces for ABI 12.

**Reply by ahlinc** [reply]
There is an issue ~#1182~ #1225 with `--prev-abi` flag, I'm not sure that the issue happen for the node.js binding but if it's true then I'd suggest to use older version of the tree-sitter CLI before 0.19.0.

**Reply by GordianDziwis** [reply]
I think you mean: https://github.com/tree-sitter/tree-sitter/issues/1225

**Reply by GordianDziwis** [reply]
@ahlinc When this issue is resolved I could fork my parser and create a separate npm package for the old abi?

**Reply by claytonrcarter** [reply]
Oh, whoops. Sorry to lead you astray! Another option you can try is to just install and use an older version of `tree-sitter-cli`:
```
$ npm install tree-sitter-cli@0.17.1
$ ./node_modules/.bin/tree-sitter generate
```

**Reply by ahlinc** [reply]
@BonaBeavis Thanks you for correcting me, I've accidentally copied other link.

**Reply by ahlinc** [reply]
@BonaBeavis I'm not sure that it will be resolved, I'm inclined to remove the flag and better put efforts on fixing issues with the actual version.

**Reply by GordianDziwis** [reply]
@ahlinc To what issue do you refer?

**Reply by ahlinc** [reply]
I meant #1225, I've tried to look on it and it seems requires too much efforts to find a bug introduced by the ABI change.

**Reply by GordianDziwis** [reply]
Okay, I will try @claytonrcarter suggestion, and just fork and create a second package which depends on an older version of treesitter.
